File: frontend/src/stores/inference.ts
import { defineStore } from 'pinia'
import { ref } from 'vue'
import type { InferenceResponse } from '@/api/types'
import {
  runPathAnalysis,
  runObservational,
  runInterventional,
  runConfounding,
  fetchInferenceHistory,
} from '@/api/inference'

export const useInferenceStore = defineStore('inference', () => {
  const result = ref<InferenceResponse | null>(null)
  const history = ref<any[]>([])
  const loading = ref(false)
  const error = ref<string | null>(null)

  async function pathAnalysis(source: string, target: string) {
    loading.value = true
    error.value = null
    try {
      result.value = await runPathAnalysis(source, target)
    } catch (e: any) {
      error.value = e.message
    } finally {
      loading.value = false
    }
  }

  async function observational(source: string, target: string, value: number) {
    loading.value = true
    error.value = null
    try {
      result.value = await runObservational(source, target, value)
    } catch (e: any) {
      error.value = e.message
    } finally {
      loading.value = false
    }
  }

  async function interventional(source: string, target: string, value: number) {
    loading.value = true
    error.value = null
    try {
      result.value = await runInterventional(source, target, value)
    } catch (e: any) {
      error.value = e.message
    } finally {
      loading.value = false
    }
  }

  async function confounding(source: string, target: string) {
    loading.value = true
    error.value = null
    try {
      result.value = await runConfounding(source, target)
    } catch (e: any) {
      error.value = e.message
    } finally {
      loading.value = false
    }
  }

  async function loadHistory(source?: string, target?: string, limit = 50) {
    try {
      history.value = await fetchInferenceHistory(source, target, limit)
    } catch (e: any) {
      error.value = e.message
    }
  }

  return { result, history, loading, error, pathAnalysis, observational, interventional, confounding, loadHistory }
})

File: api/routers/inference.py
import time
import asyncio
from functools import partial

from fastapi import APIRouter, Depends, Query, HTTPException

from api.access import require_deep_analysis
from src.storage.database import get_inference_history
from api.dependencies import get_reasoning_engine
from api.models.inference import (
    InferenceRequest,
    ObservationalRequest,
    InterventionalRequest,
    InferenceResponse,
)

router = APIRouter(prefix="/inference", tags=["inference"])


@router.post("/path-analysis", response_model=InferenceResponse)
async def path_analysis(req: InferenceRequest, user=Depends(require_deep_analysis)):
    start = time.time()
    loop = asyncio.get_event_loop()
    try:
        engine = await loop.run_in_executor(None, get_reasoning_engine)
        result = await loop.run_in_executor(
            None, partial(engine.path_analysis, req.source, req.target)
        )
    except Exception as e:
        raise HTTPException(status_code=500, detail=str(e))
    elapsed = int((time.time() - start) * 1000)
    return InferenceResponse(
        inference_type="path_analysis",
        source=req.source,
        target=req.target,
        result=result,
        computation_time_ms=elapsed,
    )


@router.post("/observational", response_model=InferenceResponse)
async def observational(req: ObservationalRequest, user=Depends(require_deep_analysis)):
    start = time.time()
    loop = asyncio.get_event_loop()
    try:
        engine = await loop.run_in_executor(None, get_reasoning_engine)
        result = await loop.run_in_executor(
            None,
            partial(engine.observational_inference, req.source, req.target, req.source_value),
        )
    except Exception as e:
        raise HTTPException(status_code=500, detail=str(e))
    elapsed = int((time.time() - start) * 1000)
    return InferenceResponse(
        inference_type="observational",
        source=req.source,
        target=req.target,
        result=result,
        computation_time_ms=elapsed,
    )


@router.post("/interventional", response_model=InferenceResponse)
async def interventional(req: InterventionalRequest, user=Depends(require_deep_analysis)):
    start = time.time()
    loop = asyncio.get_event_loop()
    try:
        engine = await loop.run_in_executor(None, get_reasoning_engine)
        result = await loop.run_in_executor(
            None,
            partial(engine.interventional_inference, req.source, req.target, req.intervention_value),
        )
    except Exception as e:
        raise HTTPException(status_code=500, detail=str(e))
    elapsed = int((time.time() - start) * 1000)
    return InferenceResponse(
        inference_type="interventional",
        source=req.source,
        target=req.target,
        result=result,
        computation_time_ms=elapsed,
    )


@router.post("/confounding", response_model=InferenceResponse)
async def confounding(req: InferenceRequest, user=Depends(require_deep_analysis)):
    start = time.time()
    loop = asyncio.get_event_loop()
    try:
        engine = await loop.run_in_executor(None, get_reasoning_engine)
        result = await loop.run_in_executor(
            None,
            partial(engine.confounding_detection, req.source, req.target),
        )
    except Exception as e:
        raise HTTPException(status_code=500, detail=str(e))
    elapsed = int((time.time() - start) * 1000)
    return InferenceResponse(
        inference_type="confounding",
        source=req.source,
        target=req.target,
        result=result,
        computation_time_ms=elapsed,
    )


@router.get("/history")
def inference_history(
    source: str = Query(default=None),
    target: str = Query(default=None),
    limit: int = Query(default=50, ge=1, le=200),
    user=Depends(require_deep_analysis),
):
    return get_inference_history(source_node=source, target_node=target, limit=limit)
